### PR TITLE
Fix variable length strings as attributes

### LINKED ISSUE #1129: Segfault when writing variable length string as attribute
```julia
julia> using HDF5

julia> fid = h5open("test.h5", "w")
🗂️ HDF5.File: (read-write) test.h5

julia> attr = create_attribute(fid, "attr-name", datatype(String), dataspace(String))
🏷️ HDF5.Attribute: attr-name

julia> write_attribute(attr, datatype(String), "attr-value")

[16750] signal (11.2): Segmentation fault: 11
in expression starting at REPL[4]:1
_platform_strlen at /usr/lib/system/libsystem_platform.dylib (unknown line)
Allocations: 1346244 (Pool: 1345285; Big: 959); GC: 2
zsh: segmentation fault  julia --project
```
using
```julia
  [f67ccb44] HDF5 v0.17.1
```
and
```julia
Julia Version 1.9.3
Commit bed2cd540a1 (2023-08-24 14:43 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 4 on 4 virtual cores
Environment:
  JULIA_VERSION = 1.9.3
  JULIA_NUM_THREADS = 4
  JULIA_PKG_SERVER_REGISTRY_PREFERENCE = eager
```

Also: the documentation around `create_attribute` is confusing. It says to use `write_attribute` to actually write the data, but if you don't use the returned `Attribute` object from `create_attribute` (instead using `write_attribute(fid, "attr-key", "attr-value")`), then it will error and say the attribute already exists.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #1129

### CODE REVIEW

File: src/attributes.jl [hunk 1]
```diff
@@ -173,8 +173,13 @@ end
 function write_attribute(attr::Attribute, memtype::Datatype, str::AbstractString)
     strbuf = Base.cconvert(Cstring, str)
     GC.@preserve strbuf begin
-        buf = Base.unsafe_convert(Ptr{UInt8}, strbuf)
-        write_attribute(attr, memtype, buf)
+        if API.h5t_is_variable_str(memtype)
+            ptr = Base.unsafe_convert(Cstring, strbuf)
+            write_attribute(attr, memtype, Ref(ptr))
+        else
+            ptr = Base.unsafe_convert(Ptr{UInt8}, strbuf)
```
Comment (mkitti) [inline src/attributes.jl:180]: Do we test this side of the conditional? I suppose this is for a fixed length string?

**Review by mkitti — COMMENTED** [review]
Where is the original test for the non-variable length side? I'm looking at this from my phone, so it is difficult to look around.

**Review by musm — APPROVED** [review]

File: src/attributes.jl (hunk 1 above)
Comment (musm) [reply src/attributes.jl:180]: Good point, maybe we should remove this. Don't think we have user-facing API that uses fixed strings

**Comment by mkitti** [comment]
Does anyone know what's going up with nightly here?

**Comment by mkitti** [comment]
```
plain: Error During Test at /home/runner/work/HDF5.jl/HDF5.jl/test/plain.jl:24
  Got exception outside of a @test
  conversion to pointer not defined for Vector{HDF5.API.hvl_t}
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:35
    [2] unsafe_convert(::Type{Ptr{HDF5.API.hvl_t}}, a::Vector{HDF5.API.hvl_t})
      @ Base ./pointer.jl:67
    [3] unsafe_convert(::Type{Ptr{Nothing}}, a::Vector{HDF5.API.hvl_t})
      @ Base ./pointer.jl:66
    [4] h5d_write(dataset_id::HDF5.Dataset, mem_type_id::HDF5.Datatype, mem_space_id::Int64, file_space_id::Int64, xfer_plist_id::HDF5.DatasetTransferProperties, buf::HDF5.VLen{HDF5.ASCIIChar})
      @ HDF5.API ~/work/HDF5.jl/HDF5.jl/src/api/functions.jl:908
    [5] write_dataset(dset::HDF5.Dataset, memtype::HDF5.Datatype, x::HDF5.VLen{HDF5.ASCIIChar}, xfer::HDF5.DatasetTransferProperties)
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:488
    [6] write_dataset(parent::HDF5.File, name::String, data::HDF5.VLen{HDF5.ASCIIChar}; pv::@Kwargs{})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:320
    [7] write_dataset(parent::HDF5.File, name::String, data::HDF5.VLen{HDF5.ASCIIChar})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:315
    [8] macro expansion
      @ ~/work/HDF5.jl/HDF5.jl/test/plain.jl:88 [inlined]
    [9] macro expansion
      @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Test/src/Test.jl:1598 [inlined]
   [10] top-level scope
      @ ~/work/HDF5.jl/HDF5.jl/test/plain.jl:27
   [11] include(fname::String)
      @ Base.MainInclude ./client.jl:497
   [12] macro expansion
      @ ~/work/HDF5.jl/HDF5.jl/test/runtests.jl:36 [inlined]
   [13] macro expansion
      @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Test/src/Test.jl:1598 [inlined]
   [14] top-level scope
      @ ~/work/HDF5.jl/HDF5.jl/test/runtests.jl:35
   [15] include(fname::String)
      @ Base.MainInclude ./client.jl:497
   [16] top-level scope
      @ none:6
   [17] eval
      @ Core ./boot.jl:418 [inlined]
   [18] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:291
   [19] _start()
      @ Base ./client.jl:568
abstract arrays: Error During Test at /home/runner/work/HDF5.jl/HDF5.jl/test/plain.jl:674
  Got exception outside of a @test
  conversion to pointer not defined for Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false}
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:35
    [2] unsafe_convert(::Type{Ptr{UInt8}}, a::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false})
      @ Base ./pointer.jl:67
    [3] unsafe_convert(::Type{Ptr{Nothing}}, a::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false})
      @ Base ./pointer.jl:66
    [4] h5d_write(dataset_id::HDF5.Dataset, mem_type_id::HDF5.Datatype, mem_space_id::Int64, file_space_id::Int64, xfer_plist_id::HDF5.DatasetTransferProperties, buf::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false})
      @ HDF5.API ~/work/HDF5.jl/HDF5.jl/src/api/functions.jl:908
    [5] write_dataset(dataset::HDF5.Dataset, memtype::HDF5.Datatype, buf::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false}, xfer::HDF5.DatasetTransferProperties)
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:523
    [6] write_dataset(dataset::HDF5.Dataset, memtype::HDF5.Datatype, buf::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:521
    [7] write_dataset(parent::HDF5.File, name::String, data::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false}; pv::@Kwargs{})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:320
    [8] write_dataset(parent::HDF5.File, name::String, data::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:315
    [9] write(parent::HDF5.File, name::String, data::Base.ReinterpretArray{UInt8, 1, Bool, Vector{Bool}, false}; pv::@Kwargs{})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/datasets.jl:342
   [10] (::var"#37#40")(f::HDF5.File)
      @ Main ~/work/HDF5.jl/HDF5.jl/test/plain.jl:681
   [11] (::HDF5.var"#17#18"{HDF5.HDF5Context, @Kwargs{}, var"#37#40", HDF5.File})()
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/file.jl:101
   [12] task_local_storage(body::HDF5.var"#17#18"{HDF5.HDF5Context, @Kwargs{}, var"#37#40", HDF5.File}, key::Symbol, val::HDF5.HDF5Context)
      @ Base ./task.jl:299
   [13] h5open(::var"#37#40", ::String, ::Vararg{String}; context::HDF5.HDF5Context, pv::@Kwargs{})
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/file.jl:96
   [14] h5open(::Function, ::String, ::String)
      @ HDF5 ~/work/HDF5.jl/HDF5.jl/src/file.jl:94
   [15] macro expansion
      @ ~/work/HDF5.jl/HDF5.jl/test/plain.jl:679 [inlined]
   [16] macro expansion
      @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Test/src/Test.jl:1598 [inlined]
   [17] top-level scope
      @ ~/work/HDF5.jl/HDF5.jl/test/plain.jl:677
   [18] include(fname::String)
      @ Base.MainInclude ./client.jl:497
   [19] macro expansion
      @ ~/work/HDF5.jl/HDF5.jl/test/runtests.jl:36 [inlined]
   [20] macro expansion
      @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Test/src/Test.jl:1598 [inlined]
   [21] top-level scope
      @ ~/work/HDF5.jl/HDF5.jl/test/runtests.jl:35
   [22] include(fname::String)
      @ Base.MainInclude ./client.jl:497
   [23] top-level scope
      @ none:6
   [24] eval
      @ Core ./boot.jl:418 [inlined]
   [25] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:[291](https://github.com/JuliaIO/HDF5.jl/actions/runs/6724886980/job/18278119057?pr=1130#step:5:294)
   [26] _start()
      @ Base ./client.jl:568
```

**Comment by ericphanson** [comment]
can a new release be tagged with this fix? 🙏

**Comment by mkitti** [comment]
I think it might be easier to back port this to the 0.17 branch. I will work on that today.

**Pull request merged**